File: 643-maximum-average-subarray-i/maximum-average-subarray-i.cpp
class Solution {
public:
    double findMaxAverage(vector<int>& nums, int k) {
        int totalSum = accumulate(nums.begin(), nums.begin() + k, 0);
        int i = 0, j = k, n = nums.size();
        double calcMean = totalSum / static_cast<double>(k);
        double maxAverage = calcMean;
        while (j < n) {
            totalSum -= nums[i++];
            totalSum += nums[j++];
            calcMean = totalSum / static_cast<double>(k);
            maxAverage = max(maxAverage, calcMean);
        }
        return maxAverage; // TC: O(n), SC: O(1)
    }
};

File: 643-maximum-average-subarray-i/maximum-average-subarray-i.java
class Solution {
    public double findMaxAverage(int[] nums, int k) {
        final int n = nums.length;
        int[] sum = new int[n];
        sum[0] = nums[0];
        for (int i = 1; i < n; i++) {
            sum[i] = sum[i-1] + nums[i];
        }
        double res = (sum[k-1] * 1.0) / k;
        for (int i = k; i < n; i++) {
            double calcMean = ((sum[i] - sum[i-k]) * 1.0) / k;
            res = Math.max(res, calcMean);
        }
        return res; // TC: O(n), SC: O(n)
    }
}